Wikipedia/Code/WelcomeAnimationExtensions.swift

```diff
@@ -0,0 +1,753 @@
+        shape.strokeColor = UIColor.blackColor().CGColor
+        shape.transform = plus.animation.initialTransform
+        shape.wmf_applyCommonSettingsForSize(toView.frame.size, position: plus.position, initialOpacity: plus.animation.initialOpacity)
+        toView.layer.addSublayer(shape)
+        return shape;
+    }
+    class func wmf_addBar(bar: WelcomeBar, toView: UIView) -> CAShapeLayer {
+        let shape = CAShapeLayer()
+        let rect = CGRectMake(
+            0.0,
+            0.0,
+            toView.frame.size.width * bar.frame.size.width,
+            toView.frame.size.height * bar.frame.size.height
+        )
+        shape.path = UIBezierPath(roundedRect: rect, cornerRadius: bar.cornerRadius).CGPath
+        shape.fillColor = bar.color.CGColor
+        shape.transform = bar.animation.initialTransform
+        shape.position = CGPointMake(
+            (toView.frame.size.width * bar.frame.origin.x),
+            (toView.frame.size.height * bar.frame.origin.y)
+        )
+        toView.layer.addSublayer(shape)
+        return shape;
+    }
+}
+
+extension CALayer{
+    func wmf_addWelcomeAnimation(animation: WelcomeAnimation) {
+        self.transform = animation.initialTransform;
+        self.addAnimation(CABasicAnimation.wmf_animationToXF(animation.finalTransform, delay: animation.delay, duration: animation.duration), forKey: nil)
+        
+        self.opacity = Float(animation.initialOpacity);
+        self.addAnimation(CABasicAnimation.wmf_animationToOpacity(animation.finalOpacity, delay: animation.delay, duration: animation.duration), forKey: nil)
+    }
+}
+
+extension CABasicAnimation{
+    class func wmf_animationToXF(transform: CATransform3D, delay: Double, duration: Double) -> CABasicAnimation {
+        let anim = CABasicAnimation(keyPath: "transform")
+        anim.duration = duration
+        anim.beginTime = CACurrentMediaTime() + delay
+        anim.fillMode = kCAFillModeForwards
+        anim.removedOnCompletion = false
+        anim.toValue = NSValue(CATransform3D: transform)
+        anim.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
+        return anim;
+    }
+    class func wmf_animationToOpacity(opacity: Double, delay: Double, duration: Double) -> CABasicAnimation {
+        let anim = CABasicAnimation(keyPath: "opacity")
+        anim.duration = duration
+        anim.beginTime = CACurrentMediaTime() + delay
+        anim.fillMode = kCAFillModeForwards
+        anim.removedOnCompletion = false
+        anim.toValue = opacity
+        anim.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
+        return anim;
+    }
+}
+
+extension UIView{
+    func wmf_addCircles(circles: WelcomeCircle...) {
+        for circle: WelcomeCircle in circles {
+            let circleShape = circle.isDashed ? CAShapeLayer.wmf_addDashedCircle(circle, toView: self) : CAShapeLayer.wmf_addCircle(circle, toView:self)
```

Comment:
A bit inside out… you made objects to encapsulate the logic for a circle, but then here you need to ask the circle a question and then decide how to draw it.

The circle should know how to draw itself. 

https://pragprog.com/articles/tell-dont-ask